### PR TITLE
api: HTTPRouteFilter

### PR DESCRIPTION
**Pull request by guydc**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Related #4120

### CODE REVIEW

File: api/v1alpha1/httpfilter_types.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	KindHTTPFilter = "HTTPFilter"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=envoy-gateway,shortName=hf
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// HTTPFilter is a custom Envoy Gateway HTTPRouteFilter which provides extended
+// traffic processing options such as path regex rewrite, direct response and more.
+type HTTPFilter struct {
```
Comment (zirain) [inline api/v1alpha1/httpfilter_types.go:23]: I'm not sure which is better, `HTTPFilter` or `HTTPRouteFilter`?

Comment (arkodg) [reply api/v1alpha1/httpfilter_types.go]: +1 for `HTTPRouteFilter`

File: api/v1alpha1/httproutefilter_types.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	// Spec defines the desired state of HTTPRouteFilter.
+	Spec HTTPRouteFilterSpec `json:"spec"`
+}
+
+// HTTPRouteFilterSpec defines the desired state of HTTPRouteFilter.
+// +union
+type HTTPRouteFilterSpec struct {
+	// +optional
+	URLRewrite HTTPURLRewriteFilter `json:"urlRewrite,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/httproutefilter_types.go:35]: should be a ptr

File: api/v1alpha1/httproutefilter_types.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	// RegexHTTPPathModifier This type of modifier indicates that the portions of the path that match the specified
+	//  regex would be substituted with the specified substitution value
+	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/regex.proto#type-matcher-v3-regexmatchandsubstitute
+	RegexHTTPPathModifier HTTPPathModifierType = "ReplaceRegexMatch"
+)
+
+type ReplaceRegexMatch struct {
+	// Regex matches a regular expression against the value of the HTTP Path.The regex string must
+	// adhere to the syntax documented in https://github.com/google/re2/wiki/Syntax.
+	Regex string `json:"regex"`
```
Comment (arkodg) [inline api/v1alpha1/httproutefilter_types.go:59]: it reads as
```
replaceRegexMatch:
  regex:
  ...
```

would the envoy field names (pattern and substitution) be better https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/regex.proto#envoy-v3-api-msg-type-matcher-v3-regexmatchandsubstitute ?
I'm just brainstorming, no strong opinion either way
emissary uses these terms and istio uses - match and rewrite

File: api/v1alpha1/httproutefilter_types.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+const (
+	// RegexHTTPPathModifier This type of modifier indicates that the portions of the path that match the specified
+	//  regex would be substituted with the specified substitution value
+	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/regex.proto#type-matcher-v3-regexmatchandsubstitute
+	RegexHTTPPathModifier HTTPPathModifierType = "ReplaceRegexMatch"
+)
+
+type ReplaceRegexMatch struct {
+	// Regex matches a regular expression against the value of the HTTP Path.The regex string must
```
Comment (arkodg) [inline api/v1alpha1/httproutefilter_types.go:57]: can we add some examples to users can better understand how capture groups work etc

File: api/v1alpha1/httproutefilter_types.go (hunk 3 above)
Comment (arkodg) [reply api/v1alpha1/httproutefilter_types.go:59]: https://gateway-api.sigs.k8s.io/geps/gep-726/#future-extension also uses `pattern` and `substitution`

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: api/v1alpha1/httproutefilter_types.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// KindHTTPRouteFilter is the name of the HTTPRouteFilter kind.
+	KindHTTPRouteFilter = "HTTPRouteFilter"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=envoy-gateway,shortName=hf
```
Comment (arkodg) [inline api/v1alpha1/httproutefilter_types.go:18]: hrf ?

Comment (guydc) [reply api/v1alpha1/httproutefilter_types.go:18]: yes!

**Review by arkodg — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Pull request merged**